File: examples/nodejs/query_format.mjs
#!/usr/bin/env node
/**
 * Run a literal SELECT on the Spanner emulator and format cells with spanvalue.
 */

import { Spanner } from '@google-cloud/spanner';
import {
  adaptClientType,
  encodeValue,
  formatResultRow,
  simpleFormatConfig,
} from '@apstndb/spanvalue';

const SQL = "SELECT 1 AS n, 'hello' AS s, true AS b";

process.env.SPANNER_EMULATOR_HOST ??= 'localhost:9010';

const projectId = process.env.SPANNER_PROJECT_ID ?? 'test-project';
const instanceId = process.env.SPANNER_INSTANCE_ID ?? 'test-instance';
const databaseId = process.env.SPANNER_DATABASE_ID ?? 'test-db';

const spanner = new Spanner({ projectId });
const database = spanner.instance(instanceId).database(databaseId);
const config = simpleFormatConfig();

const [rows, , response] = await database.run({ sql: SQL });
if (rows.length === 0) {
  console.error('Query returned no rows.');
  process.exit(1);
}

const fields = response.metadata.rowType.fields;
const colTypes = fields.map((field) => adaptClientType(field.type));
const row = rows[0];
const nativeValues = fields.map((_, index) => row[index]);

const wireValue = encodeValue(colTypes[0], nativeValues[0]);
console.log(`encodeValue (n): ${JSON.stringify(wireValue)}`);

const formatted = formatResultRow(colTypes, nativeValues, config);
console.log(`formatResultRow: ${JSON.stringify(formatted)}`);

File: examples/python/query_format.py
#!/usr/bin/env python3
"""Run a literal SELECT on the Spanner emulator and format cells with spanvalue."""

from __future__ import annotations

import os
import sys

os.environ.setdefault("SPANNER_EMULATOR_HOST", "localhost:9010")

from google.cloud import spanner  # noqa: E402

from spanvalue import (  # noqa: E402
    adapt_client_type,
    encode_value,
    format_result_row,
    simple_format_config,
)

SQL = "SELECT 1 AS n, 'hello' AS s, true AS b"


def main() -> int:
    project_id = os.environ.get("SPANNER_PROJECT_ID", "test-project")
    instance_id = os.environ.get("SPANNER_INSTANCE_ID", "test-instance")
    database_id = os.environ.get("SPANNER_DATABASE_ID", "test-db")

    client = spanner.Client(project=project_id)
    database = client.instance(instance_id).database(database_id)
    config = simple_format_config()

    with database.snapshot() as snapshot:
        result_set = snapshot.execute_sql(SQL)
        fields = result_set.metadata.row_type.fields
        col_types = [adapt_client_type(field.type) for field in fields]

        for row in result_set:
            native_values = list(row)

            # Full path for column 0: metadata type → adapt → encode → format cell
            wire_type = col_types[0]
            wire_value = encode_value(wire_type, native_values[0])
            print(f"encode_value (n): {wire_value}")

            formatted = format_result_row(col_types, native_values, config)
            print(f"format_result_row: {formatted}")
            return 0

    print("Query returned no rows.", file=sys.stderr)
    return 1


if __name__ == "__main__":
    raise SystemExit(main())
